### PR TITLE
SINGA-347 Create a function that supports einsum

### PR DESCRIPTION
**Pull request by Sheyujian**
The High dimension matrix matrix/vector calculation

### CODE REVIEW

File: include/singa/core/tensor.h [hunk 1]
```diff
@@ -440,7 +440,14 @@ void Mult(const Tensor &A, const Tensor &B, Tensor *C);
 template <typename SType>
 void Mult(const SType alpha, const Tensor &A, const Tensor &B, const SType beta,
           Tensor *C);
-
+/// Do high dimension matrix multipication on the tensor shape. result = A * B
```
Comment (nudles) [inline include/singa/core/tensor.h]: what's the difference of HighDimMult and GEMM or GEMV?

File: python/singa/tensor.py [hunk 2]
```diff
@@ -938,7 +938,69 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
         singa.MultWithScale(alpha, A.singa_tensor, B.singa_tensor,
                             beta, C.singa_tensor)
         return C
-
+def tensormult (A,B,axes=2):
+    try:
```
Comment (nudles) [inline python/singa/tensor.py]: pls add some comments for this function. it is difficult to understand the logic.

File: src/core/tensor/tensor.cc [hunk 3]
```diff
@@ -1050,6 +1050,54 @@ void Mult(const SType alpha, const Tensor &A, const Tensor &B, const SType beta,
...
+    return out;
+}
+    
+void HighDimMult(const Tensor &A, const Tensor &B, Tensor *out) {
+    HighDimMult(1.0f, A, B, 0.0f, out);
+}
+    
+template <typename SType>
+void HighDimMult(const SType alpha, const Tensor &A, const Tensor &B, const SType beta, Tensor *C) {
+    CHECK_GE(A.shape().size(), 3u);
```
Comment (nudles) [inline src/core/tensor/tensor.cc]: tensordot can be implemented using GEMM and GEMV?

File: src/core/tensor/tensor.cc [hunk 4]
```diff
@@ -1050,6 +1050,54 @@ void Mult(const SType alpha, const Tensor &A, const Tensor &B, const SType beta,
...
+                GEMM<DType, Lang>(A_2d.transpose(), B_2d.transpose(), A_2d.shape(0), B_2d.shape(1),
+                                    A_2d.shape(1), a, A_2d.block(), B_2d.block(), b, C->block(),ctx);
+            }, {A_2d.block(), B_2d.block()}, {C->block()});
+        });
+    }
+}
+//Tensor TensorMult(const Tensor &A,const Tensor &B, int axes)
 // ************************
 // Misc.
 // ************************
```
Comment (nudles) [inline src/core/tensor/tensor.cc]: remove .DS_Store files before you push to github.

File: include/singa/core/tensor.h (hunk 1 above)
Comment (SYJhuangtao) [reply include/singa/core/tensor.h]: The GEMM and GEMV(which use cblas_sgemm or cblas_sgemm) could only calculate the 2D matrix or vector multiplication. When the matrix becomes 3D or above, here I use the HighDimMult to do the multiplication.

File: python/singa/tensor.py [hunk 5]
```diff
@@ -938,7 +938,104 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+    Return:
+        singa.tensor: The tensor  product of ''a'' and ''b'' along the
+        axes specified by ''axes''.
+        """
+    # when axes is an integer, axes_A and axes_B represent axes at the last of ''a`'' and
+    # the first of ''b''. For example, when axes is 1, we do the normal multiplication :
+    # if A is in shape(3,2,4), B is in shape(4,2,5), it will return a matrix in shape(3,2,2,5)
+    #when axes is 2 and A,B are in the same shape, it will return a matrix in shape(3,5)
+    try:
+        iter(axes)
```
Comment (nudles) [inline python/singa/tensor.py]: why not check the type directly?

`if type(axes) == int`

File: python/singa/tensor.py [hunk 6]
```diff
@@ -938,7 +938,104 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+    # nda is the dim of A, and axes_a is the axis for A, notin is the axis which is not in axes_A
+    newaxes_a = notin + axes_a
+    N2 = 1
+    for axis in axes_a:
+        N2 *= a_shape[axis]
+    N1 = 1
+    for ax in notin:
+        N1 *=a_shape[ax]
+    # newshape_a is the shape to do multiplication.For example, A is in shape(3,2,4),
+    #B is in shape(4,2,5), we set axes as ([1,2],[1,0]), then newshape_a should be (3,8)
```
Comment (nudles) [inline python/singa/tensor.py]: newshape_a should be (3, 5) or (3, 8)?

File: python/singa/tensor.py [hunk 7]
```diff
@@ -938,7 +938,104 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+        iter(axes)
+    except Exception:
+        axes_A = list(range(-axes, 0))
+        axes_B = list(range(0, axes))
+    else:
+        axes_A, axes_B = axes
+    # when axes is a pair of sequences of integers.For example, A is in shape(3,2,4),
+    #B is in shape(4,2,5), we set axes as ([1,2],[1,0]), it will return a matrix in shape(3,5)
+    try:
+        na = len(axes_A)
```
Comment (nudles) [inline python/singa/tensor.py]: again, you can check 

`type(axes_A) == list` or `isintance(axes_A, list)`

File: python/singa/tensor.py [hunk 8]
```diff
@@ -938,7 +938,104 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+        axes_A = list(axes_B)
```
Comment (nudles) [inline python/singa/tensor.py]: axes_B or axes_A?

File: python/singa/tensor.py [hunk 9]
```diff
@@ -938,7 +938,104 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+    # to check if the length of axe_A is equal to axes_B
+    if na != nb:
+        equal = False
+    else:
+    # to make the shape match
+        for k in range(na):
+            if a_shape[axes_a[k]] != b_shape[axes_b[k]]:
+                equal = False
+                break
+            if axes_a[k] < 0:
```
Comment (nudles) [inline python/singa/tensor.py]: axes_a or axes_A?

File: python/singa/tensor.py [hunk 10]
```diff
@@ -940,6 +940,179 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+    at = a.transpose(newaxes_a).reshape(newshape_a)
+    bt = b.transpose(newaxes_b).reshape(newshape_b)
+    res = mult(at, bt)
+    # reshape the result
+    return res.reshape(olda + oldb)
+
+
+def einsum(A,B,ops):
+    '''
+    Thanks to nils-werner/sparse.einsum()
```
Comment (nudles) [inline python/singa/tensor.py]: again make it explicit. I cannot find the source based on nils-werner/sparse.einsum()

File: python/singa/tensor.py [hunk 11]
```diff
@@ -940,6 +940,179 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+    sums = [[inop.find(x) for x in sums] for inop in inputops]
+    # Find output axes in input axes for final transpose
+    # Values very likely lie outside of output tensor shape, so
+    # just map them values to their rank (index in ordered list)
+    transpose = [''.join(inputops).find(x) for x in outputops]
+    transpose = sorted(range(len(transpose)), key = transpose.__getitem__)
+
+    return tensormult(A,B, sum=sums, multiply=multiplies).transpose(transpose)
+
+def tensordotmult(A, B, sum=None, multiply=None):
```
Comment (nudles) [inline python/singa/tensor.py]: what's the difference of tensormult and tensordot?

File: python/singa/tensor.py [hunk 12]
```diff
@@ -940,6 +940,309 @@ def mult(A, B, C=None, alpha=1.0, beta=0.0):
...
+    for i in index:
+        i_diag = list(i)
+        i_diag.insert(axis, i[axis])
+        index_npA = [[i[x]] for x in range(len(i))]
+        index_npA_diag = [[i_diag[x]] for x in range(len(i_diag))]
+        npA_diag[index_npA_diag] = npA[index_npA]
+    A_diag = from_numpy(npA_diag)
+    return A_diag
+
+def einsum2(A,B,ops):
```
Comment (nudles) [inline python/singa/tensor.py]: the previous implementation can be removed?
add  a TODO list to include the numpy functions to be replaced by singa functions.

Comment (Sheyujian) [reply python/singa/tensor.py]: the einsum2 could realize the what the einsum do.  I will remove the previous einsum and add the TODO list.

**Pull request merged**